### PR TITLE
Warn on unused `offset_of!()` result

### LINKED ISSUE #111669: Not using the result of offset_of should lint
The offset_of macro (#106655) is side effect free. For standard library functions that are side effect free, or that have side effects but there is versions of them that don't return anything, it is policy to add a `must_use` attribute to them so that users are warned if they invoke the function without doing anything with the result (#89692).
 
Ideally, the same should hold for the `offset_of` macro as well. One of the lints in the `unused` lint group should fire if you don't use the result of the `offset_of!()` macro:
```Rust
fn foo() {
    offset_of!(F, field); // WARNING result must be used
}
struct F { field: u8, }
```

Ideally it would be integrated into the `must_use` system, say via supporting `#[must_use]` attributes on macros. Currently you are getting:

```
warning: `#[must_use]` has no effect when applied to a macro def
 --> src/main.rs:1:1
  |
1 | #[must_use = "hi"]
  | ^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default
```
for
```Rust
#[must_use = "hi"]
macro_rules! foo {
    () => {0};
}
fn main() {
    foo!();
}
```

It could e.g. expand to some AST node that is just the identity.

Alternatively, one could also just solve it for the offset_of macro via sending the result through a function:

```Rust
macro_rules! offset_of_must_use {
    () => {
            #[must_use = "offset_of result must be used"]
            const fn must_use(v: usize) -> usize { v }
            must_use(0)
    };
}

fn main() {
    offset_of_must_use!();
}
```

Sadly this pollutes the context with a function however. If one surrounds the entire expansion with `{}`s, the lint is silent again.

I don't think it's worth its own lint in itself.

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
The usage of `core::hint::must_use()` means that we don't get a specialized message. I figured out that since there are plenty of other methods that just have `#[must_use]` with no message it'll be fine, but it is a bit unfortunate that the error mentions `must_use` and not `offset_of!`.

Fixes #111669.

### CODE REVIEW

File: library/core/src/mem/mod.rs [hunk 1]
```diff
@@ -1317,7 +1317,7 @@ impl<T> SizedTypeProperties for T {}
 /// ```
 #[cfg(not(bootstrap))]
 #[unstable(feature = "offset_of", issue = "106655")]
-#[allow_internal_unstable(builtin_syntax)]
+#[allow_internal_unstable(builtin_syntax, hint_must_use)]
 pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
-    builtin # offset_of($Container, $($fields).+)
+    crate::hint::must_use(builtin # offset_of($Container, $($fields).+))
```
Comment (est31) [inline library/core/src/mem/mod.rs:1322]: I think this needs a `$` at the front

Comment (ChayimFriedman2) [reply library/core/src/mem/mod.rs:1322]: No it doesn't because this is a macros 2.0 macro, it is resolved at macro's definition.

Comment (est31) [reply library/core/src/mem/mod.rs:1322]: It is? That's cool.

File: tests/ui/offset-of/offset-of-arg-count.rs [hunk 2]
```diff
@@ -9,7 +9,9 @@ fn main() {
     offset_of!(S, f); // compiles fine
     offset_of!(S, f,); // also compiles fine
     offset_of!(S, f.); //~ ERROR unexpected end of macro invocation
-    offset_of!(S, f.,); //~ ERROR expected identifier
+    offset_of!(S, f.,);
+    //~^ ERROR expected identifier
+    //~| ERROR expected one of `.`, `;`, `?`, `}`, or an operator, found `)`
```
[the remark this answers is not in the document]
Comment (est31) [reply tests/ui/offset-of/offset-of-arg-count.rs:14]: One could try maybe surrounding the builtin with `{}`s like: `crate::hint::must_use({builtin # offset_of($Container, $($fields).+)})`. What is the error then?

[the remark this answers is not in the document]
Comment (ChayimFriedman2) [reply tests/ui/offset-of/offset-of-arg-count.rs]: @est31 It worked!

**Comment by WaffleLapkin** [comment]
@ChayimFriedman2 you should probably rebase on master and then run `./x.py test ests/ui/offset-of` to see which tests fail because of those changes and fix them (by adding `let _ = `s, I suppose).

File: tests/ui/macros/offset-of-must-use.rs
Comment (WaffleLapkin) [inline tests/ui/macros/offset-of-must-use.rs:1]: also I'd probably move this to `ui/offset-of` dir

File: tests/ui/offset-of/offset-of-output-type.stderr [hunk 3]
```diff
@@ -2,49 +2,39 @@ error[E0308]: mismatched types
   --> $DIR/offset-of-output-type.rs:12:17
    |
 LL |     let _: u8 = offset_of!(S, v);
-   |            --   ^^^^^^^^^^^^^^^^ expected `u8`, found `usize`
-   |            |
-   |            expected due to this
```
Comment (WaffleLapkin) [inline tests/ui/offset-of/offset-of-output-type.stderr:7]: Kind of a shame we are missing those, but I'm not sure what to do with that...

**Review by WaffleLapkin — APPROVED** [review]

**Comment by est31** [comment]
@ChayimFriedman2 fixing this should be easy, just rebase the PR, ignoring any merge conflicts, and then do:

```
./x.py test tests/mir-opt --bless
```

**Comment by ChayimFriedman2** [comment]
`--bless` for some reason did not work, I have no idea why but it outputted incorrect results... I had to do that manually.

**Comment by ChayimFriedman2** [comment]
I don't understand. This is exactly the opposite of the previous failure. Something strange is happening here.

**Comment by ChayimFriedman2** [comment]
If I does what bors wants (which BTW is what `--bless` emits), then it fails locally. What should I do?

**Comment by ChayimFriedman2** [comment]
@oli-obk 

> what's your host?

x86_64-pc-windows-msvc

> do you have a `config.toml`? If so, what did you configure?

It is configured to `profile = "compiler"`.

> is it only failing for the 32 bit target but works for the 64 bit target?

No, it fails for the 64 bit target (and doesn't check 32 bit therefore).

**Comment by oli-obk** [comment]
wait I'm confused. If you `--bless`, then the tests fail locally? Can you gist your entire log of that happening?

**Comment by ChayimFriedman2** [comment]
@oli-obk 

```
Building bootstrap
    Finished dev [unoptimized] target(s) in 0.19s
Building stage0 library artifacts (x86_64-pc-windows-msvc)
    Finished release [optimized] target(s) in 0.54s
Building compiler artifacts (stage0 -> stage1, x86_64-pc-windows-msvc)
    Finished release [optimized] target(s) in 1.66s
Creating a sysroot for stage1 compiler (use `rustup toolchain link 'name' build/host/stage1`)
Building stage1 library artifacts (x86_64-pc-windows-msvc)
    Finished release [optimized] target(s) in 1.01s
Building stage0 tool compiletest (x86_64-pc-windows-msvc)
    Finished release [optimized] target(s) in 0.69s
Check compiletest suite=mir-opt mode=mir-opt (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)

running 1 tests
F

failures:

---- [mir-opt] tests\mir-opt\const_prop\offset_of.rs stdout ----
29                StorageLive(_2);
30      -         _2 = OffsetOf(Alpha, [0]);
31      +         _2 = const 4_usize;
-                 _1 = must_use::<usize>(move _2) -> [return: bb1, unwind unreachable];
+                 _1 = must_use::<usize>(move _2) -> bb1;
33            }
34
35            bb1: {

38                StorageLive(_4);
39      -         _4 = OffsetOf(Alpha, [1]);
40      +         _4 = const 0_usize;
-                 _3 = must_use::<usize>(move _4) -> [return: bb2, unwind unreachable];
+                 _3 = must_use::<usize>(move _4) -> bb2;
42            }
43
44            bb2: {

47                StorageLive(_6);
48      -         _6 = OffsetOf(Alpha, [2, 0]);
49      +         _6 = const 2_usize;
-                 _5 = must_use::<usize>(move _6) -> [return: bb3, unwind unreachable];
+                 _5 = must_use::<usize>(move _6) -> bb3;
51            }
52
53            bb3: {

56                StorageLive(_8);
57      -         _8 = OffsetOf(Alpha, [2, 1]);
58      +         _8 = const 3_usize;
-                 _7 = must_use::<usize>(move _8) -> [return: bb4, unwind unreachable];
+                 _7 = must_use::<usize>(move _8) -> bb4;
60            }
61
62            bb4: {

thread '[mir-opt] tests\mir-opt\const_prop\offset_of.rs' panicked at 'Actual MIR output differs from expected MIR output C:\rust\tests\mir-opt\const_prop\offset_of.concrete.ConstProp.diff', src\tools\compiletest\src\runtest.rs:3652:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    [mir-opt] tests\mir-opt\const_prop\offset_of.rs

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 260 filtered out; finished in 891.43ms
```

**Comment by ChayimFriedman2** [comment]
That is, `--bless` adds the possibility of unwinding, but locally it wants it to be non-unwinding call (probably it determines that the function cannot unwind). On borts, https://github.com/rust-lang/rust/pull/111684#issuecomment-1597898289 wants unwinding, but previously https://github.com/rust-lang/rust/pull/111684#issuecomment-1596206453 wanted non-unwinding.

**Comment by oli-obk** [comment]
🤦 your test needs a `// EMIT_MIR_FOR_EACH_PANIC_STRATEGY` comment at the top of the file so that we end up with two `diff` files

**Pull request merged**